botwo/botwo.py

```diff
@@ -0,0 +1,200 @@
+                res = self._route_params(kwargs["CopySource"])
+                client_to_call_source = res[0]
+                kwargs["CopySource"] = res[1]
+
+            if kwargs.get("Bucket"):
+                res = self._route_params(kwargs)
+                client_to_call_dest = res[0]
+                api_params = res[1]
+
+            if client_to_call_source != client_to_call_dest:
+                raise ValueError("client source and client destination are different")
+
+            return getattr(client_to_call_source, operation_name)(**api_params)
+
+        _api_call.__name__ = str(operation_name)
+        return _api_call
+
+    def _create_get_paginate_method(self, operation_name):
+        def _paginator_api_call(*args, **kwargs):
+            return PaginatorWrapper(self.mapping, self.config, kwargs)
+
+        _paginator_api_call.__name__ = str(operation_name)
+        return _paginator_api_call
+
+    def _create_can_paginate_method(self, operation_name):
+        def _can_paginate_api_call(*args, **kwargs):
+            return getattr(self.default, operation_name)(**kwargs)
+
+        _can_paginate_api_call.__name__ = str(operation_name)
+        return _can_paginate_api_call
+
+    def _create_delete_objects_method(self, operation_name):
+        def _delete_objects_api_call(_, *args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            if kwargs.get("Delete"):  # delete objects operation
+                for i, obj in enumerate(kwargs["Delete"]["Objects"]):
+                    params = {"Bucket": kwargs.get("Bucket"), "Key": obj["Key"]}
+                    res = self._route_params(params)
+                    client_to_call = res[0]
+                    bucket = res[1]["Bucket"]
+                    kwargs["Delete"]["Objects"][i]["Key"] = res[1]["Key"]
+                    # TODO (eden) fail if client to call not equal to previous
+            kwargs["Bucket"] = bucket
+            return getattr(client_to_call, operation_name)(**kwargs)
+
+        _delete_objects_api_call.__name__ = str(operation_name)
+        return _delete_objects_api_call
+
+    def _route_params(self, api_params):
+        for profile in self.config:
+            mapping = self.config[profile]
+
+            if fnmatch.fnmatch(api_params["Bucket"], mapping.get("source_bucket_pattern")):
+                if api_params["Key"]:
+                    if mapping.get("source_key_pattern"):
+                        if not fnmatch.fnmatch(api_params["Key"], mapping.get("source_key_pattern")):
+                            continue
+                    if mapping.get("mapped_prefix"):
+                        api_params["Key"] = mapping.get("mapped_prefix") + api_params["Key"]
+                if mapping.get("mapped_bucket_name"):
+                    api_params["Bucket"] = mapping.get("mapped_bucket_name")
```

Comment:
All this code is basically duplicated in the paginator.

This function can be `_route_bucket_and_key(bucket, key)`.
Then you can use it in the paginator as is.
Also then the usage in line 117 will be more natural instead of artificially creating a map to pass down to this function.